src/Share/Codebase.hs

```diff
@@ -185,20 +182,21 @@ codebaseEnv !_authZReceipt codebaseLoc = do
   let codebaseOwner = Codebase.codebaseOwnerUserId codebaseLoc
    in CodebaseEnv {codebaseOwner}
 
--- | Construct a Runtime linked to a specific codebase.
+-- | Construct a Runtime linked to a specific codebase and transaction.
 -- Don't use the runtime for one codebase with another codebase.
-codebaseRuntime :: (MonadReader (Env.Env x) m, MonadUnliftIO m, Env.HasTags x) => CodebaseEnv -> m (CodebaseRuntime IO)
-codebaseRuntime codebase = do
-  unisonRuntime <- asks Env.sandboxedRuntime
-  rt <- liftIO (codebaseRuntimeTransaction unisonRuntime codebase)
-  unlift <- badAskUnliftCodebaseRuntime
-  pure (unlift rt)
+-- Don't use this runtime in any transaction other than the one where it's created.
+withCodebaseRuntime :: (Exception e) => CodebaseEnv -> Runtime Symbol -> (forall s. CodebaseRuntime s IO -> PG.Transaction e r) -> PG.Transaction e r
```

Comment:
It looks like this would have the same issues as `badAskUnliftCodebaseRuntime`, but in fact it actually will use the same connection as the Transaction this is run within, so it avoids the possibility of connection claiming deadlock.